### PR TITLE
Add option for truncated stream detection

### PR DESCRIPTION
**Pull request by mfkl**
Having another go at https://github.com/dotnet/runtime/issues/47563 (previous https://github.com/dotnet/runtime/pull/61768)

This time using an opt-in flag to avoid breakage like this https://github.com/dotnet/runtime/issues/72726.

/cc @Jozkee

### CODE REVIEW

File: src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs [hunk 1]
```diff
@@ -479,6 +478,8 @@ async Task<long> GetLengthAsync(CompressionLevel compressionLevel)
         [InlineData(TestScenario.ReadByteAsync)]
         public async Task StreamTruncation_IsDetected(TestScenario scenario)
         {
+            AppContext.SetSwitch("System.IO.Compression.UseStrictValidation", true);
```
Comment (stephentoub) [inline src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs:481]: This affects all tests process-wide. Any test that changes such a process-wide switch needs to use RemoteExecutor to move that code out of process.

File: src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs [hunk 2]
```diff
@@ -146,6 +151,42 @@ public static void StreamsEqual(Stream ast, Stream bst, int blocksToRead)
...
+
+                ac = await ReadAllBytesAsync(ast, ad, 0, 4096);
+                bc = await ReadAllBytesAsync(bst, bd, 0, 4096);
+
+                if (ac != bc)
+                {
+                    bd = NormalizeLineEndings(bd);
+                }
+
+                Assert.True(ArraysEqual<byte>(ad, bd, ac), "Stream contents not equal: " + ast.ToString() + ", " + bst.ToString());
```
Comment (stephentoub) [inline src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs:184]: This should use AssertExtensions.SequenceEqual.

File: src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs [hunk 3]
```diff
@@ -227,9 +234,23 @@ private bool TryDecompress(Span<byte> destination, out int bytesWritten)
...
+        private static bool GetStrictValidationSetting()
+        {
+            if (AppContext.TryGetSwitch("System.IO.Compression.UseStrictValidation", out bool strictValidation))
+            {
+                return strictValidation;
+            }
+
+            string? envVar = Environment.GetEnvironmentVariable("DOTNET_SYSTEM_COMPRESSION_STRICT_VALIDATION");
+            return envVar is not null && (envVar == "1" || envVar.Equals("true", StringComparison.OrdinalIgnoreCase));
+        }
```
Comment (stephentoub) [inline src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs:246]: This should be cached into a static readonly bool.

File: src/libraries/System.IO.Compression/src/Resources/Strings.resx [hunk 4]
```diff
@@ -278,6 +278,9 @@
   <data name="SplitSpanned" xml:space="preserve">
     <value>Split or spanned archives are not supported.</value>
   </data>
+  <data name="TruncatedData" xml:space="preserve">
+    <value>Found truncated data while decoding.</value>
```
Comment (stephentoub) [inline src/libraries/System.IO.Compression/src/Resources/Strings.resx:282]: The text of the same error message should be the same across libs.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs [hunk 5]
```diff
@@ -1049,5 +1078,16 @@ private void AsyncOperationStarting()
...
+        private static bool GetStrictValidationSetting()
+        {
+            if (AppContext.TryGetSwitch("System.IO.Compression.UseStrictValidation", out bool strictValidation))
+            {
+                return strictValidation;
+            }
+
+            string? envVar = Environment.GetEnvironmentVariable("DOTNET_SYSTEM_COMPRESSION_STRICT_VALIDATION");
+            return envVar is not null && (envVar == "1" || envVar.Equals("true", StringComparison.OrdinalIgnoreCase));
+        }
```
Comment (stephentoub) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:1091]: As in the other library, this needs to be cached into a static readonly bool.

File: src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs (hunk 1 above)
Comment (mfkl) [reply src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs]: Ok, done.

File: src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs [hunk 6]
```diff
@@ -146,6 +151,42 @@ public static void StreamsEqual(Stream ast, Stream bst, int blocksToRead)
...
+            int blocksRead = 0;
+
+            //assume read doesn't do weird things
+            do
+            {
+                if (blocksToRead != -1 && blocksRead >= blocksToRead)
+                    break;
+
+                ac = await ReadAllBytesAsync(ast, ad, 0, 4096);
+                bc = await ReadAllBytesAsync(bst, bd, 0, 4096);
```
[the remark this answers is not in the document]
Comment (mfkl) [reply src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs]: Ok

File: src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs (hunk 2 above)
Comment (mfkl) [reply src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs]: Done

File: src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs (hunk 3 above)
Comment (mfkl) [reply src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs]: I considered that but thought we wanted to be able to change that during runtime. Guess not. Changed.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 5 above)
Comment (mfkl) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs]: Yes, changed.

File: src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs [hunk 7]
```diff
@@ -146,6 +151,42 @@ public static void StreamsEqual(Stream ast, Stream bst, int blocksToRead)
...
+            int blocksRead = 0;
+
+            //assume read doesn't do weird things
+            do
+            {
+                if (blocksToRead != -1 && blocksRead >= blocksToRead)
+                    break;
+
+                ac = await ast.ReadAtLeastAsync(ad, 4096);
+                bc = await bst.ReadAtLeastAsync(bd, 4096);
```
Comment (stephentoub) [inline src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs:177]: Presumably you want to pass throwOnEndOfStream:false for both of these?

File: src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs [hunk 8]
```diff
@@ -146,6 +151,42 @@ public static void StreamsEqual(Stream ast, Stream bst, int blocksToRead)
...
+
+                if (ac != bc)
+                {
+                    bd = NormalizeLineEndings(bd);
+                }
+
+                AssertExtensions.SequenceEqual(ad, bd);
```
Comment (stephentoub) [inline src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs:184]: And these should be sliced, e.g. `ad.AsSpan(0, ac), bd.AsSpan(0, bc)`.

If this is just copying what's already there for the sync implementation, then maybe inputs just aren't triggering the faulty behavior?

File: src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs [hunk 9]
```diff
@@ -227,9 +234,15 @@ private bool TryDecompress(Span<byte> destination, out int bytesWritten)
             return false;
         }
 
+        private static readonly bool UseStrictValidation =
```
Comment (stephentoub) [inline src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs:237]: UseStrictValidation => s_useStrictValidation

Comment (mfkl) [reply src/libraries/System.IO.Compression.Brotli/src/System/IO/Compression/dec/BrotliStream.Decompress.cs]: Ok

File: src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs (hunk 7 above)
Comment (mfkl) [reply src/libraries/Common/tests/System/IO/Compression/ZipTestHelper.cs]: Yes.. thanks.

**Comment by mfkl** [comment]
CI is fixed. 

Feel free to let me know what you prefer regarding https://github.com/dotnet/runtime/pull/75671#discussion_r977370461 and if anyone has any other feedback regarding this PR.

File: src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.ZLib.cs [hunk 10]
```diff
@@ -19,44 +21,134 @@ public class ZLibStreamUnitTests : CompressionStreamUnitTestBase
...
+        }
+
+        [InlineData(TestScenario.ReadAsync)]
+        [InlineData(TestScenario.Read)]
+        [InlineData(TestScenario.Copy)]
+        [InlineData(TestScenario.CopyAsync)]
+        [InlineData(TestScenario.ReadByte)]
+        [InlineData(TestScenario.ReadByteAsync)]
+        [ConditionalTheory(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
+        public void StreamTruncation_IsDetected(TestScenario testScenario)
```
Comment (Jozkee) [inline src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.ZLib.cs:75]: This test is duplicated for Deflate, Gzip and Zlib tests. Why did you move it out of `CompressionStreamUnitTestBase`?

File: src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs (hunk 1 above)
Comment (Jozkee) [reply src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs]: Another option is to create a second test project that enables the appcontext switch and runs all tests of the "normal" project with the switch on.

Comment (Jozkee) [reply src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs]: And a third option could be to turn the switch on for all the tests, we already know our tests will pass with the "strict validation" in place.

File: src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs [hunk 11]
```diff
@@ -275,7 +277,6 @@ public void DerivedStream_ReadWriteSpan_UsesReadWriteArray()
         }
 
 
-        [ActiveIssue("https://github.com/dotnet/runtime/issues/47563")]
         [Fact]
         public void StreamCorruption_IsDetected()
```
Comment (Jozkee) [inline src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs:281]: This test also needed the RemoteExecutor + enable the switch, no? I am a bit surprised that CI didn't fail.

File: src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.ZLib.cs (hunk 10 above)
Comment (mfkl) [reply src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.ZLib.cs:75]: https://github.com/dotnet/runtime/pull/75671/commits/fa4b613a9c3fe4f13d8b2fc147676ca2d2e92236

> abstracted classes are not supported by RemoteExecutor

File: src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs (hunk 11 above)
Comment (mfkl) [reply src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs:281]: Gzip detects the corruption without the strict validation enabled (thanks to its CRC integrity check, I assume). 

zlib not so much, if I remove the `AppContext.SetSwitch("System.IO.Compression.UseStrictValidation", true)`, the test fails.

File: src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs (hunk 1 above)
Comment (mfkl) [reply src/libraries/Common/tests/System/IO/Compression/CompressionStreamUnitTestBase.cs]: > a third option could be to turn the switch on for all the tests, we already know our tests will pass with the "strict validation" in place.

Not anymore, this one will fail https://github.com/dotnet/runtime/blob/eead714e4a5e82f2075de7d5eaad75dedf509941/src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Deflate.cs#L78

> Another option is to create a second test project that enables the appcontext switch and runs all tests of the "normal" project with the switch on.

I prefer the current way (with the RemoteExecutor) but no strong feelings. Let me know.

**Comment by mfkl** [comment]
I believe I've addressed most concerns. Do you have any more feedback on this PR?

**Review by jozkee — APPROVED** [review]
LGTM, thanks.

CI issue looks unrelated to this PR.
